src/shared/sol-iio.c

```diff
@@ -0,0 +1,990 @@
+    }
+
+    if (!write_file(path, "%s", trigger_name)) {
+        SOL_WRN("Could not write to device current_trigger file");
+        return false;
+    }
+
+    return true;
+}
+
+static bool
+create_sysfs_trigger(struct sol_iio_device *device)
+{
+    int len, i;
+    struct stat st;
+    char *trigger_name;
+    bool result = false;
+
+    /* First, check if 'iio_sysfs_trigger' is available */
+    if (stat(SYSFS_TRIGGER_SYSFS_ADD_TRIGGER, &st)) {
+        SOL_WRN("No 'iio_sysfs_tigger' under '/sys/bus/iio/devices'."
+            " Missing 'modprobe iio-trig-sysfs'?");
+        return false;
+    }
+
+    /* TODO check available trigger indice instead of using a random one */
+    /* Create new trigger */
+    i = rand();
+    if (!write_file(SYSFS_TRIGGER_SYSFS_ADD_TRIGGER, "%d", i))
+        goto error;
+
+    /* Set device current trigger */
+    len = asprintf(&trigger_name, "sysfstrig%d", i);
+    if (len < 0)
+        goto error;
+    result = set_current_trigger(device, trigger_name);
+
+    if (result)
+        device->trigger_name = strdup(trigger_name);
+
+    free(trigger_name);
+
+    return result;
+
+error:
+    SOL_WRN("Could not create sysfs trigger.");
+
+    return false;
+}
+
+static bool
+check_trigger(struct sol_iio_device *device)
+{
+    char path[PATH_MAX + 1];
+    char *name = NULL;
+    int len;
+
+    len = snprintf(path, sizeof(path), CURRENT_TRIGGER_DEVICE_PATH, device->device_id);
+    if (len < 0 || len > PATH_MAX) {
+        SOL_WRN("Could not open IIO device%d trigger on sysfs", device->device_id);
+        return false;
+    }
+
+    len = sol_util_read_file(path, "%ms", &name);
+    if (len < 0) {
+        SOL_INF("No current trigger for iio:device%d. Creating a sysfs one.",
+            device->device_id);
+        if (!create_sysfs_trigger(device))
+            return false;
+    } else
+        device->trigger_name = name;
+
+    return check_manual_triggering(device);
+}
+
+static void
+set_buffer_size(struct sol_iio_device *device, int buffer_size)
+{
+    char path[PATH_MAX + 1];
+
+    if (craft_filename_path(path, sizeof(path), BUFFER_LENGHT_DEVICE_PATH, device->device_id)) {
+        SOL_WRN("Could not set IIO device buffer size");
+        return;
+    }
+
+    if (!write_file(path, "%d", buffer_size)) {
+        SOL_WRN("Could not set IIO device buffer size");
+    }
+
+    return;
+}
+
+static bool
+set_buffer_enabled(struct sol_iio_device *device, bool enabled)
+{
+    char path[PATH_MAX + 1];
+    int len;
+
+    len = snprintf(path, sizeof(path), BUFFER_ENABLE_DEVICE_PATH, device->device_id);
+    if (len < 0 || len > PATH_MAX)
+        return false;
+
+    if (!write_file(path, "%d", enabled))
+        return false;
+
+    return true;
+}
+
+static size_t
+calc_buffer_size(struct sol_iio_device *device)
+{
+    struct sol_iio_channel *channel;
+    int i;
+    size_t size = 0;
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&device->channels, channel, i) {
+        size += channel->storagebits;
```

Comment:
`storagebits` should be the total number of bits that a channel use. `bits` is the one without paddings.